contracts/mirror_staking/src/staking.rs

```diff
@@ -148,6 +152,134 @@ pub fn decrease_short_token<S: Storage, A: Api, Q: Querier>(
+    for asset in assets.iter() {
+        match asset.info.clone() {
+            AssetInfo::NativeToken { .. } => {
+                asset.assert_sent_native_token_balance(&env)?;
+                native_asset_op = Some(asset.clone())
+            }
+            AssetInfo::Token { contract_addr } => {
+                token_info_op = Some((contract_addr, asset.amount))
+            }
+        }
+    }
+
+    // will fail if one of them is missing
+    let native_asset: Asset = match native_asset_op {
+        Some(v) => v,
+        None => return Err(StdError::generic_err("Missing native asset")),
+    };
+    let (token_addr, token_amount) = match token_info_op {
+        Some(v) => v,
+        None => return Err(StdError::generic_err("Missing token asset")),
+    };
+
+    // query pair info to obtain pair contract address
+    let asset_infos: [AssetInfo; 2] = [assets[0].info.clone(), assets[1].info.clone()];
+    let terraswap_pair: PairInfo = query_pair_info(deps, &terraswap_factory, &asset_infos)?;
+
+    // compute tax
+    let tax_amount: Uint128 = native_asset.compute_tax(deps)?;
+
+    // 1. Transfer token asset to staking contract
+    // 2. Increase allowance of token for pair contract
+    // 3. Provide liquidity
+    // 4. Execute staking hook, will stake in the name of the sender
+    Ok(HandleResponse {
+        messages: vec![
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: token_addr.clone(),
+                msg: to_binary(&Cw20HandleMsg::TransferFrom {
+                    owner: env.message.sender.clone(),
+                    recipient: env.contract.address.clone(),
+                    amount: token_amount,
+                })?,
+                send: vec![],
+            }),
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: token_addr.clone(),
+                msg: to_binary(&Cw20HandleMsg::IncreaseAllowance {
+                    spender: terraswap_pair.contract_addr.clone(),
+                    amount: token_amount,
+                    expires: None,
+                })?,
+                send: vec![],
+            }),
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: terraswap_pair.contract_addr,
+                msg: to_binary(&PairHandleMsg::ProvideLiquidity {
+                    assets: [
+                        Asset {
+                            amount: (native_asset.amount.clone() - tax_amount)?,
+                            info: native_asset.info.clone(),
+                        },
+                        Asset {
+                            amount: token_amount,
+                            info: AssetInfo::Token {
+                                contract_addr: token_addr.clone(),
+                            },
+                        },
+                    ],
+                    slippage_tolerance,
+                })?,
+                send: vec![native_asset.deduct_tax(deps)?],
```

Comment:
`deduct_tax` conduct two queries `tax_rate` & `tax_cap`.
We can skip this step by subtracting the pre-computed tax_amount like L224.

Reply:
changed it!